### PR TITLE
feat: add profile routing and guards

### PR DESCRIPTION
**Pull request by ttalhouk**
Set up profile page guard and protection for nested edit page. 

Profile landing page can only be accessed to logged in user.
Logged in user can access any profile via /profile/:id (eventually remove the edit button if current user !== params.id)
Logged in user can only access their own edit page (beforeEnter guard) via /profile/:id/edit otherwise redirected to /profile/:id

So far only a skeleton set up, as I want to make sure the routing is how we want it before we start adding functionality.   Should the sub components such as ProfileDetails and ProfileEdit go in a separate directory?

### CODE REVIEW

File: src/pages/Profile.vue [hunk 1]
```diff
@@ -1,6 +1,8 @@
 <template>
   <div>
     <h1>Profile Page</h1>
+    <router-link :to="'/profile/' + $auth.currentUser.uid ">My Profile</router-link>
```
Comment (shaun-sweet) [inline src/pages/Profile.vue]: lets use the fancy template literals!

```js
`/profile/${$auth.currentUser.uid}`
```

File: src/router/index.js [hunk 2]
```diff
@@ -38,8 +40,27 @@ let router = new Router({
...
+      meta: {
+        requiresAuth: true,
+      },
+      children: [
+        { path: ':id', component: ProfileDetail },
+        {
+          path: ':id/edit',
+          component: ProfileEdit,
+          beforeEnter (to, from, next) {
+            const routeId = to.params.id
```
Comment (shaun-sweet) [inline src/router/index.js]: this is the userId yah? perhaps can name it userId vs routeId

File: src/router/index.js [hunk 3]
```diff
@@ -38,8 +40,27 @@ let router = new Router({
...
+            const currentUser = Auth.auth.currentUser
+            if (routeId !== currentUser.uid) {
+              next(`/profile/${routeId}`)
+            } else {
+              next()
```
Comment (shaun-sweet) [inline src/router/index.js]: so i'm a bit of a fan of returning next() type middleware stuff even though you don't need the return value, i think it reads a bit better than if/else.  The returns just serve to break the function 

~~also, where is `Auth` coming from?~~

File: src/router/index.js [hunk 4]
```diff
@@ -55,7 +76,7 @@ let router = new Router({
 })
 
 router.beforeEach((to, from, next) => {
-  let currentUser = Auth.auth.currentUser
+  const currentUser = Auth.auth.currentUser
   let requiresAuth = to.matched.some(record => record.meta.requiresAuth)
```
Comment (shaun-sweet) [inline src/router/index.js]: i think this can be const as well if i'm reading right.

**Review by shaun-sweet — CHANGES_REQUESTED** [review]
I think a higher level question i had while going over this is this... What are you envisioning the differences between the Profile and ProfileDetail?

So I'm not sure about the directories.  Personally I like them flat because its easier to import stuff as needed and I just hit command + P in the editor and type what i'm looking for to bring up whatever file (as opposed to browsing a file tree)

File: src/pages/Profile.vue (hunk 1 above)
Comment (shaun-sweet) [reply src/pages/Profile.vue]: whether we use the state for the comparing UID vs auth shouldn't matter, right?

Comment (ttalhouk) [reply src/pages/Profile.vue]: Aggreed I would prefer to use template literals, but my highlighting wasn't working.  Do I need to put the whole thing in {} or just drop the template literals directly into the :to

**Comment by ttalhouk** [comment]
So right now profile page template holds the router-view for the child components. So if you visit profile/:id it pulls up the profile page and throws the profile details in the router-view.  If you click edit and are allowed to edit that perticular profile it loads the ':id/edit' child route in the router view within profile.  So basically profile is a container for all the child routes and could contain links to profile related things then the specific route would appear under that.  

Alternately we could just render each route to the main router in which the profile would be equivilent to profileDetails.

File: src/pages/Profile.vue (hunk 1 above)
Comment (ttalhouk) [reply src/pages/Profile.vue]: The store uid should be the same as the $auth.currentUser.uid so I could change that to the store getter for the uid.

File: src/router/index.js (hunk 2 above)
Comment (ttalhouk) [reply src/router/index.js]: True but it is the userId in the route, as opposed to the userId of the current user which could be different if someone is trying to edit someone elses profile.  Just tried to keep them seperate as much as possible but I'm not stuck to this naming convention so if you have any suggestions I can go with it.

File: src/pages/Profile.vue (hunk 1 above)
Comment (shaun-sweet) [reply src/pages/Profile.vue]: i don't really care where we get the uid from, mostly just wanna make sure whatever we decide is consistent as much as possible

Comment (shaun-sweet) [reply src/pages/Profile.vue]: so i use vscode and the highlighting is awesome for vue once you install vetur plugin.  i had a hard time getting the HTML portion to syntax highlight properly in atom.  maybe we should make a vue plugin for atom to fix that lol.  

i'm not entirely sure.  I would imagine you can just drop the entire back tick template in.  Anything in the quote is valid JS i believe.  You could also make a computed property for it and dump that in.

**Pull request merged**